src/lib/build/job.rs

```diff
@@ -215,22 +217,33 @@ impl JobQueue {
                             // there's no lib target and nothing to depend on.
                             self.graph[job_index].work = Work::None
                         }
-                        
-                        let mut parents_done = true;
 
                         // push jobs that can be execute at next step into queue
                         for (parent, _) in self.graph.parents(job_index) {
-                            parents_done &= !self.graph[parent].is_dirty();
-                            let ready = self.graph.children(parent).all(|(_, job)| job.is_fresh());
+                            let ready = self
+                                .graph
+                                .children(parent)
+                                .all(|(_, job)| job.work.is_fresh());
 
-                            if ready && self.graph[parent].is_dirty() {
+                            if ready && self.graph[parent].work.is_dirty() {
                                 queue.push(parent);
                             }
                         }
-                        
-                        // If all of the parents of this dep are done, we can set ourselves to None
-                        if parents_done {
-                            self.graph[job_index].work = Work::None;
+
+                        // If the parents of any of the childs are done, we can set it to None
+                        let mut child_done = Vec::new();
```

Comment:
We should check the childs but not the node itself; the parent is not going to be built before this node is finished.